x-pack/test/detection_engine_api_integration/utils/wait_for_rule_status.ts

```diff
@@ -7,33 +7,51 @@
 
 import type { ToolingLog } from '@kbn/tooling-log';
 import type SuperTest from 'supertest';
-
 import { DETECTION_ENGINE_RULES_URL } from '@kbn/security-solution-plugin/common/constants';
 import { RuleExecutionStatus } from '@kbn/security-solution-plugin/common/detection_engine/rule_monitoring';
 import { waitFor } from './wait_for';
 
+interface WaitForRuleStatusBaseParams {
+  supertest: SuperTest.SuperTest<SuperTest.Test>;
+  log: ToolingLog;
+  afterDate?: Date;
+}
+
+interface WaitForRuleStatusWithId extends WaitForRuleStatusBaseParams {
+  id: string;
+}
+
+interface WaitForRuleStatusWithRuleId extends WaitForRuleStatusBaseParams {
+  ruleId: string;
+}
```

Comment:
If I get your idea correctly, for these types to become discriminated unions, you need to change them to:

```ts
interface WaitForRuleStatusWithId extends WaitForRuleStatusBaseParams {
  id: string;
  ruleId?: undefined;
}

interface WaitForRuleStatusWithRuleId extends WaitForRuleStatusBaseParams {
  id?: undefined;
  ruleId: string;
}
```

That way, either `id` or `ruleId` will have to be provided to `waitForRuleStatus`, but not both of them at the same time.